feat: add metadata to hive ui with linked directive support

jdolle opened the pull request:
### Background

We need a way to provide and visualize metadata for subgraphs.

https://github.com/graphql-hive/console/issues/2747
https://github.com/graphql-hive/console/issues/2818

<!---
Please include a short note explaining the need for this PR / change.
If you are resolving/closing/fixing an issue, please mention it in this section.
--->

### Description

This adds metadata to fields via a directive.
In this PR, I added a way to view metadata on fields via the explorer. But to close these issues, we will need additional views and behaviors. E.g. filtering the explorer based on metadata, and viewing metadata on a field's insights page.
<img width="812" alt="Screenshot 2025-02-13 at 12 54 30 PM" src="https://github.com/user-attachments/assets/13fae6ad-676d-4f92-ba75-479f1d1ebea9" />



<!---
Please share here a technical description of your changes. This should include what packages/components are effects: CLI, client/agent, services, APIs.
--->

### Checklist

<!---
We are following the OWASP Secure Coding Practices for develpoing Hive. You can find the complete guide here:
https://owasp.org/www-pdf-archive/OWASP_SCP_Quick_Reference_Guide_v2.pdf

Please use this checklist to ensure your PR quality before proceeding.
You may remove unnecessary checks from this list, if it's not relevant to your changes.
--->

- [x] Testing

jdolle commented on packages/libraries/federation-link-utils/src/index.ts:
```diff
@@ -46,12 +46,6 @@ export function extractLinkImplementations(typeDefs: DocumentNode): {
-      if (!supportsFederationV2) {
-        // Identities dont matter for Federation v1. There are no links to reference.
-        // So return the name without the identity's namespace
-        return name.startsWith('@') ? name.substring(1) : name;
```
Im hitting issues with this logic. It isnt completely accurate...
To know for sure I need to add a check for the `@core` directive also.
WIP

jdolle commented:
I've adjusted this to save metadata to a column on the schemaVersion table.
Since we're capable of parsing the SDL at runtime, looking up the metadata in an object should be no worse.

jdolle commented on packages/services/api/src/modules/schema/resolvers/SchemaMetadata.ts:
```diff
@@ -0,0 +1,14 @@
+ *
+ * If you want to skip this file generation, remove the mapper or update the pattern in the `resolverGeneration.object` config.
+ */
+export const SchemaMetadata: SchemaMetadataResolvers = {
+  /* Implement SchemaMetadata resolver logic here */
+};
```
identity resolver is correct in this case. There's no need to implement anything.

jdolle commented on packages/libraries/cli/examples/federation.products.graphql:
```diff
@@ -1,60 +1,74 @@
-enum CURRENCY_CODE {
```
I assume the prior version was for Federation 1? But I couldn't get the two example schemas working locally. I've opted to update them to Federation 2. Let me know if I should make a new version instead.

dotansimha commented on integration-tests/tests/schema/metadata.spec.ts:
```diff
@@ -0,0 +1,152 @@
+import type { SchemaBuilderApi } from '@hive/schema';
+import { createTRPCProxyClient, httpLink } from '@trpc/client';
+
+const host =
```
nit: should this go somewhere into the test kit? 😄

jdolle replied:
possibly... I dont see anywhere in the codebase where we use `SCHEMA_SERVICE_HOST_OVERRIDE`... If anything I assume `RUN_AGAINST_LOCAL_SERVICES` is more practical for us to use for overriding the hosts.
I'd be in favor of removing the env var condition in this file and in the `contracts.spec.ts` before moving this into the test kit.

jdolle commented on packages/migrations/src/actions/2025.02.14T00-00-00.schema-versions-metadata.ts:
```diff
@@ -0,0 +1,10 @@
+import type { MigrationExecutor } from '../pg-migrator';
+
+export default {
+  name: '2025.02.14T00.00.00.schema-versions-metadata.ts',
```
No one would be using `@meta` -- since we're not just looking for that directive here, but we're also making sure it's linked to our spec url

dotansimha replied:
great!

n1ru4l commented on packages/services/schema/src/lib/metadata-extraction.ts:
Let's have some tests for this implementation.

jdolle replied:
Covered via the integration test. I'm happy to add unit tests too if preferred.

n1ru4l commented on packages/services/schema/src/orchestrators.ts:
```diff
@@ -214,6 +277,16 @@ const createFederation: (
+          try {
+            const subgraphMetadatas = subgraphs.map(({ typeDefs }) => extractMetadata(typeDefs));
+            metadata = mergeMetadata(...subgraphMetadatas);
+          } catch (e: unknown) {
+            // warn and proceed because metadata isn't considered critical
+            logger.warn(`Cannot extract metadata from subgraphs because ${e}`);
+          }
```
What could be thrown there?

jdolle replied:
This is a remnant of a previous iteration. Ive removed the try/catch. It shouldn't throw.

n1ru4l commented on packages/services/schema/src/orchestrators.ts:
```diff
@@ -273,6 +348,7 @@ const createFederation: (
               filter,
             );
             return {
+              // @note Although it can differ from the supergraph's, ignore metadata on contracts.
               ...subgraph,
               typeDefs: filteredSubgraph.typeDefs,
```
What is the reason here for ignoring the metadata?

jdolle replied:
Metadata will only differ between contracts and supergraph for `@shared` fields or entities, where one subgraph could impact another.
I think this is an edge case and the additional complexity of calculating and displaying metadata doesn't provide much benefit to the users.

n1ru4l reviewed (approved):
Overall, I think the implementation looks solid to me. ✅

I would like some more unit or integration tests around the metadata extraction to have a bit more confidence and make sure we are not breaking something unexpectedly.

I want @kamilkisiela to review this as well before merging things, to double check this covers what our clients need.

jdolle commented on packages/migrations/test/2024.07.23T09.36.00.schema-cleanup-tracker.test.ts:
```diff
@@ -93,7 +93,8 @@ await describe('migration: schema-cleanup-tracker', async () => {
             has_contract_composition_errors,
-            conditional_breaking_change_metadata
+            conditional_breaking_change_metadata,
+            schemaMetadata,
```
yup!

jdolle commented:
> Could there be a use case where people want to specify meta-data but not have it appear within the supergraph?

`@composeDirective` was required in the first iteration, but isn't anymore. Sorry, I forgot to update the test.

I've updated the tests to cover more cases and removed the `@composeDirective` to clarify that it isnt required.

kamilkisiela commented:
I will review the code tomorrow, but in terms of UI, I think we could pass the metadata to the subgraph badge.
Metadata is bound to a subgraph anyway, and when hovering on the subgraph, we could show extra information (like contact details).

If at least two subgraphs share a field and those team add metadata (like slack channel or just the name of the team), a dedicated icon for metadata will fall short. It won't be able to display info about two teams. If we merge it into the subgraph badge, it connects nicely.

jdolle commented:
That's an interesting point. I like the idea of linking it to the subgraph, and I agree that fewer icons the better.
I'll make that adjustment.

jdolle commented:
<img width="863" alt="Screenshot 2025-02-19 at 4 23 33 PM" src="https://github.com/user-attachments/assets/7cbd588c-8d9d-40e6-9f40-f33b647b92f4" />
I've updated the API to add an additional "source" field to indicate the origin of the metadata.
I felt "source" was more accurate than "subgraph" if ever this metadata can be generated in another way.
